physics-informed-neural-networks-for-steady-cavity-flow/README.md

```diff
@@ -0,0 +1,274 @@
+
+# Cavity flow with Physics\-Informed Neural Networks
+
+Solve cavity flow governed by 2d steady Navier\-Stokes equations and continuity equation, using a Physics\-Informed Neural Network (PINN).
+
+
+The 2d, steady Navier\-Stokes equations for an incompressible fluid are:
+
+ $$ \frac{\partial u}{\partial x}+\frac{\partial v}{\partial y}=0 $$ 
+
+ $$ u\frac{\partial u}{\partial x}+v\frac{\partial u}{\partial y}+\frac{\partial p}{\partial x}-\frac{1}{Re}\bigg(\frac{\partial^2 u}{\partial x^2 }+\frac{\partial^2 u}{\partial y^2 }\bigg)=0 $$ 
+
+ $$ u\frac{\partial v}{\partial x}+v\frac{\partial v}{\partial y}+\frac{\partial p}{\partial y}-\frac{1}{Re}\bigg(\frac{\partial^2 v}{\partial x^2 }+\frac{\partial^2 v}{\partial y^2 }\bigg)=0 $$ 
+
+`(x,y)` are the spatial coordinates, `(u,v)` is the fluid velocity, `p` is the pressure, `Re` is the Reynolds number.
+
+
+In order to automatically satisfy the continuity equation we use the stream function psi such that `u=psi_y` and `v=-psi_x`. The boundary conditions are `(u,v)=(1,0)` and the top boundary and `(u,v)=(0,0)` at the other boundaries. The Reynolds number `Re=100`.
+
+
+The PINNs model takes the spatial coordinates `(x,y)` as inputs and returns the streamfunction and pressure `(psi,p)` as outputs.
+
+# Set parameters.
+```matlab
+Re = 100;
+u0 = 1;
+```
+# Create network
+```matlab
+% Create basic MLP network architecture with two inputs (x,y) and two
+% outputs (psi,p).
+numHiddenUnits = 32;
+net = dlnetwork();
+layers = [ featureInputLayer(1, Name="x")
+    concatenationLayer(1, 2)
+    fullyConnectedLayer(numHiddenUnits)
+    swishLayer()
+    fullyConnectedLayer(numHiddenUnits)
+    swishLayer()
+    fullyConnectedLayer(numHiddenUnits)
+    swishLayer() 
+    fullyConnectedLayer(numHiddenUnits)
+    swishLayer(Name="swishout")
+    fullyConnectedLayer(1, Name="psiFree") ];
+net = addLayers(net, layers);
+net = addLayers(net, fullyConnectedLayer(1, Name="p"));
+net = connectLayers(net, "swishout", "p");
+net = addInputLayer(net, featureInputLayer(1, Name="y"), Initialize=false);
+
+% Add anchor functions to strictly enforce boundary conditions on the
+% streamfunction.
+net = addLayers(net, [functionLayer(@(x)4.*x.*(1-x), Name="anchorX", Acceleratable=true); multiplicationLayer(3, Name="psi")]);
+net = addLayers(net, functionLayer(@(y)4.*y.*(1-y), Name="anchorY", Acceleratable=true));
+net = connectLayers(net, "x", "anchorX");
+net = connectLayers(net, "y", "anchorY");
+net = connectLayers(net, "anchorY", "psi/in2");
+net = connectLayers(net, "psiFree", "psi/in3");
+
+% Make sure outputs are ordered (psi,p).
+net.OutputNames = ["psi", "p"];
+
+% Initialize the network and cast to double precision.
+net = initialize(net);
+net = dlupdate(@double, net);
```

Comment:
Can we add a plot of the network to the readme? I think it'd help make it clearer how the anchors enforce the boundary conditions.

Reply:
Good idea -- I'll add that.